File: dialogplus/src/test/java/com/orhanobut/android/dialogplus/ListHolderTest.java
package com.orhanobut.android.dialogplus;

import android.app.Activity;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.LinearLayout;
import android.widget.ListView;

import com.orhanobut.dialogplus.BuildConfig;
import com.orhanobut.dialogplus.HolderAdapter;
import com.orhanobut.dialogplus.ListHolder;
import com.orhanobut.dialogplus.OnHolderListener;
import com.orhanobut.dialogplus.R;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricGradleTestRunner;
import org.robolectric.annotation.Config;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.fail;

@RunWith(RobolectricGradleTestRunner.class)
@Config(constants = BuildConfig.class, sdk = 21)
public class ListHolderTest {

  Context context;

  @Before public void setup() {
    context = Robolectric.setupActivity(Activity.class);
  }

  private ListHolder getListHolder() {
    ListHolder holder = new ListHolder();
    LayoutInflater layoutInflater = LayoutInflater.from(context);
    holder.getView(layoutInflater, new LinearLayout(context));
    return holder;
  }

  @Test public void init() {
    assertThat(getListHolder()).isInstanceOf(HolderAdapter.class);
    assertThat(getListHolder()).isNotNull();
  }

  @Test public void testViewInflation() {
    ListHolder holder = new ListHolder();
    LayoutInflater layoutInflater = LayoutInflater.from(context);
    View view = holder.getView(layoutInflater, new LinearLayout(context));

    assertThat(view).isNotNull();
    assertThat(holder.getInflatedView().getId()).isEqualTo(R.id.dialogplus_list);

    ListView listView = (ListView) holder.getInflatedView();
    assertThat(listView.getOnItemClickListener()).isInstanceOf(ListHolder.class);
  }

  @Test public void testFooter() {
    ListHolder holder = getListHolder();

    assertThat(holder.getFooter()).isNull();

    View footer = new LinearLayout(context);
    holder.addFooter(footer);

    assertThat(holder.getFooter()).isEqualTo(footer);
  }

  @Test public void testHeader() {
    ListHolder holder = getListHolder();

    assertThat(holder.getHeader()).isNull();

    View header = new LinearLayout(context);
    holder.addHeader(header);

    assertThat(holder.getHeader()).isEqualTo(header);
  }

  @Test public void testOnItemClickWithoutItemListenerAndAdapter() {
    ListHolder holder = getListHolder();
    ListView listView = (ListView) holder.getInflatedView();

    try {
      listView.performItemClick(null, 0, 0);
    } catch (Exception e) {
      fail("it should not crash");
    }
  }


  @Test public void testOnItemClickWithoutItemListenerOnly() {
    ListHolder holder = getListHolder();
    ListView listView = (ListView) holder.getInflatedView();

    //with adapter set
    ArrayAdapter<String> adapter = new ArrayAdapter<>(
        context, android.R.layout.simple_list_item_1,
        new String[]{"test"}
    );
    holder.setAdapter(adapter);
    try {
      listView.performItemClick(null, 0, 0);
    } catch (Exception e) {
      fail("it should not crash");
    }
  }

  @Test public void testOnItemClick() {
    ListHolder holder = getListHolder();
    ListView listView = (ListView) holder.getInflatedView();

    //with adapter set
    ArrayAdapter<String> adapter = new ArrayAdapter<>(
        context, android.R.layout.simple_list_item_1,
        new String[]{"test"}
    );
    holder.setAdapter(adapter);

    //set listener
    holder.setOnItemClickListener(new OnHolderListener() {
      @Override
      public void onItemClick(Object item, View view, int position) {
        assertThat(String.valueOf(item)).isEqualTo("test");
        assertThat(position).isEqualTo(0);
        assertThat(view).isNull();
      }
    });
    listView.performItemClick(null, 0, 0);
  }

  @Test public void doNotCountHeaderForPositionCalculation() {
    ListHolder holder = getListHolder();
    holder.addHeader(new View(context));
    ListView listView = (ListView) holder.getInflatedView();

    //with adapter set
    ArrayAdapter<String> adapter = new ArrayAdapter<>(
        context, android.R.layout.simple_list_item_1,
        new String[]{"test"}
    );
    holder.setAdapter(adapter);

    //set listener
    holder.setOnItemClickListener(new OnHolderListener() {
      @Override
      public void onItemClick(Object item, View view, int position) {
        assertThat(String.valueOf(item)).isEqualTo("test");
        assertThat(position).isEqualTo(0);
        assertThat(view).isNull();
      }
    });
    listView.performItemClick(null, 1, 0);
  }

}

File: dialogplus/src/test/java/com/orhanobut/android/dialogplus/ListHolderTest.kt
package com.orhanobut.android.dialogplus

import android.app.Activity
import android.view.LayoutInflater
import android.view.View
import android.widget.ArrayAdapter
import android.widget.LinearLayout
import android.widget.ListView
import com.orhanobut.dialogplus.HolderAdapter
import com.orhanobut.dialogplus.ListHolder
import com.orhanobut.dialogplus.R
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.fail
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class ListHolderTest {

  private val context = Robolectric.setupActivity(Activity::class.java)

  private val listHolder: ListHolder
    get() {
      val holder = ListHolder()
      val layoutInflater = LayoutInflater.from(context)
      holder.getView(layoutInflater, LinearLayout(context))
      return holder
    }

  @Test fun init() {
    assertThat(listHolder).isInstanceOf(HolderAdapter::class.java)
    assertThat(listHolder).isNotNull()
  }

  @Test fun testViewInflation() {
    val holder = ListHolder()
    val layoutInflater = LayoutInflater.from(context)
    val view = holder.getView(layoutInflater, LinearLayout(context))

    assertThat(view).isNotNull()
    assertThat(holder.inflatedView.id).isEqualTo(R.id.dialogplus_list)

    val listView = holder.inflatedView as ListView
    assertThat(listView.onItemClickListener).isInstanceOf(ListHolder::class.java)
  }

  @Test fun testFooter() {
    val holder = listHolder

    assertThat(holder.footer).isNull()

    val footer = LinearLayout(context)
    holder.addFooter(footer)

    assertThat(holder.footer).isEqualTo(footer)
  }

  @Test fun testHeader() {
    val holder = listHolder

    assertThat(holder.header).isNull()

    val header = LinearLayout(context)
    holder.addHeader(header)

    assertThat(holder.header).isEqualTo(header)
  }

  @Test fun testOnItemClickWithoutItemListenerAndAdapter() {
    val holder = listHolder
    val listView = holder.inflatedView as ListView

    try {
      listView.performItemClick(null, 0, 0)
    } catch (e: Exception) {
      fail("it should not crash")
    }

  }


  @Test fun testOnItemClickWithoutItemListenerOnly() {
    val holder = listHolder
    val listView = holder.inflatedView as ListView

    //with adapter set
    val adapter = ArrayAdapter(
        context, android.R.layout.simple_list_item_1,
        arrayOf("test")
    )
    holder.setAdapter(adapter)
    try {
      listView.performItemClick(null, 0, 0)
    } catch (e: Exception) {
      fail("it should not crash")
    }

  }

  @Test fun testOnItemClick() {
    val holder = listHolder
    val listView = holder.inflatedView as ListView

    //with adapter set
    val adapter = ArrayAdapter(
        context, android.R.layout.simple_list_item_1,
        arrayOf("test")
    )
    holder.setAdapter(adapter)

    //set listener
    holder.setOnItemClickListener { item, view, position ->
      assertThat(item.toString()).isEqualTo("test")
      assertThat(position).isEqualTo(0)
      assertThat(view).isEqualTo(listView)
    }
    listView.performItemClick(listView, 0, 0)
  }

  @Test fun doNotCountHeaderForPositionCalculation() {
    val holder = listHolder
    holder.addHeader(View(context))
    val listView = holder.inflatedView as ListView

    //with adapter set
    val adapter = ArrayAdapter(
        context, android.R.layout.simple_list_item_1,
        arrayOf("test")
    )
    holder.setAdapter(adapter)

    //set listener
    holder.setOnItemClickListener { item, view, position ->
      assertThat(item.toString()).isEqualTo("test")
      assertThat(position).isEqualTo(0)
      assertThat(view).isEqualTo(listView)
    }
    listView.performItemClick(listView, 1, 0)
  }

}
